### PR TITLE
`link-to-changelog-file`: Exclude non-textual files

### LINKED ISSUE #4577: False positive in `link-to-changelog-file`, linking to a `.js` file which is not the changelog
### Please ensure:

- [X] The bug is caused by Refined GitHub. It doesn't happen if I disable the extension.

### Describe the problem and how to replicate it

RGH adds the `CHANGELOG` button on the top of the [releases page of hyper](https://github.com/vercel/hyper/releases), with a link to the https://github.com/vercel/hyper/blob/HEAD/release.js page, which is not the changelog (I'm actually not sure whether there even is one, though...)

### Screenshots/video/gif showing the issue, if it’s visual

grafik

### Console errors, if any

_No response_

### Example URL

https://github.com/vercel/hyper/releases

### Browser(s) used

Firefox 90.0 (64-Bit)

### PR DESCRIPTION
**Pull request by darkred**
Closes #4577 (following https://github.com/sindresorhus/refined-github/issues/4577#issuecomment-881951763)

## Test URLs

Match: (as before)
https://github.com/pyca/cryptography/releases (`CHANGELOG.rst`)
https://github.com/nodeca/js-yaml/releases/tag/4.0.0  (`CHANGELOG.md`)

Not match: 
https://github.com/vercel/hyper/releases  (`release.js`)

## Screenshot

(none)

## Notes to reviewers

 - The markdown extension regex is from: [`vertical-front-matter`](https://github.com/sindresorhus/refined-github/blob/31a04dd8b43d4fe121bf5a467dbc0cdada568756/source/features/vertical-front-matter.tsx#L9) / [`view-markdown-source`](https://github.com/sindresorhus/refined-github/blob/31a04dd8b43d4fe121bf5a467dbc0cdada568756/source/features/view-markdown-source.tsx#L34)
    ```regex
    /\.(mdx?|mkdn?|mdwn|mdown|markdown|litcoffee)$/
    ```
    I added to it: `|txt|rst` to maintain matches from the previous related issue examples.

- I noticed that `findChangelogName()` was initially returning string or false, that's why the lengthy:

    ```js
    function findChangelogName(files: string[]): string | false {
    	const filename = files.find(name => regexChangelogMarkdownTextFiles.test(name))!;
    	const result = (typeof filename === 'undefined') ? false : filename;

    	return result;
    }
    ```

    Otherwise, if returning `false` can be omitted, I could replace it with:

    ```js
    function findChangelogName(files: string[]): string{
    	return files.find(name => regexChangelogMarkdownTextFiles.test(name))!;
    }
    ```

### CODE REVIEW

File: source/features/link-to-changelog-file.tsx [hunk 1]
```diff
@@ -15,16 +15,13 @@ interface FileType {
...
 
+const regexChangelogMarkdownTextFiles = /^(changelog|news|changes|history|release|whatsnew)\.(mdx?|mkdn?|mdwn|mdown|markdown|litcoffee|txt|rst)$/i;
 function findChangelogName(files: string[]): string | false {
-	for (const file of files) {
-		if (changelogNames.has(file.toLowerCase().split('.', 1)[0])) {
-			return file;
-		}
-	}
+	const filename = files.find(name => regexChangelogMarkdownTextFiles.test(name))!;
+	const result = (typeof filename === 'undefined') ? false : filename;
```
Comment (cheap-glitch) [inline source/features/link-to-changelog-file.tsx:22]:
```suggestion
	return filename ?? false;
```

File: source/features/link-to-changelog-file.tsx [hunk 2]
```diff
@@ -15,16 +15,13 @@ interface FileType {
...
-const changelogNames = new Set(['changelog', 'news', 'changes', 'history', 'release', 'whatsnew']);
 
+const regexChangelogMarkdownTextFiles = /^(changelog|news|changes|history|release|whatsnew)\.(mdx?|mkdn?|mdwn|mdown|markdown|litcoffee|txt|rst)$/i;
 function findChangelogName(files: string[]): string | false {
-	for (const file of files) {
-		if (changelogNames.has(file.toLowerCase().split('.', 1)[0])) {
-			return file;
-		}
-	}
+	const filename = files.find(name => regexChangelogMarkdownTextFiles.test(name))!;
```
Comment (cheap-glitch) [inline source/features/link-to-changelog-file.tsx:21]: I don't think you need the assertion here:
```suggestion
	const filename = files.find(name => regexChangelogMarkdownTextFiles.test(name));
```

File: source/features/link-to-changelog-file.tsx [hunk 3]
```diff
@@ -15,16 +15,13 @@ interface FileType {
 }
 
 const getCacheKey = (): string => `changelog:${getRepo()!.nameWithOwner}`;
-const changelogNames = new Set(['changelog', 'news', 'changes', 'history', 'release', 'whatsnew']);
 
+const regexChangelogMarkdownTextFiles = /^(changelog|news|changes|history|release|whatsnew)\.(mdx?|mkdn?|mdwn|mdown|markdown|litcoffee|txt|rst)$/i;
```
Comment (cheap-glitch) [inline source/features/link-to-changelog-file.tsx:19]: There are changelogs with no extension (e.g. https://github.com/sphinx-doc/sphinx/blob/4.x/CHANGES)
```suggestion
const regexChangelogMarkdownTextFiles = /^(changelog|news|changes|history|release|whatsnew)(\.(mdx?|mkdn?|mdwn|mdown|markdown|litcoffee|txt|rst))?$/i;
```

Also the name could be shorter, it's not always Markdown and we know it's a regex by looking at it. Maybe just `changelogFiles`?

**Review by cheap-glitch — CHANGES_REQUESTED** [review]

File: source/features/link-to-changelog-file.tsx (hunk 3 above)
Comment (darkred) [reply source/features/link-to-changelog-file.tsx]: > it's not always Markdown

You've read my mind! I was going to ask about files that have no extension 👍 

> Maybe just `changelogFiles`?

Yes, 👍

**Review by cheap-glitch — APPROVED** [review]
Tested this with all the links from #4005 and #4071 :+1: :100:

File: source/features/link-to-changelog-file.tsx [hunk 4]
```diff
@@ -15,16 +15,11 @@ interface FileType {
...
 function findChangelogName(files: string[]): string | false {
-	for (const file of files) {
-		if (changelogNames.has(file.toLowerCase().split('.', 1)[0])) {
-			return file;
-		}
-	}
-
-	return false;
+	const filename = files.find(name => changelogFiles.test(name));
+	return filename ?? false;
```
Comment (fregante) [inline source/features/link-to-changelog-file.tsx:22]: Let’s just change the function’s signature to return `string | undefined` so you can skip this intermediary variable and extra line

**Review by fregante — APPROVED** [review]

File: source/features/link-to-changelog-file.tsx (hunk 4 above)
Comment (darkred) [reply source/features/link-to-changelog-file.tsx]: Done in the new commit.
I also provided an empty string as fallback value `?? ''` in the two function calls.
It should be ok.

**Review by yakov116 — APPROVED** [review]
LGTM

File: source/features/link-to-changelog-file.tsx (hunk 4 above)
Comment (cheap-glitch) [reply source/features/link-to-changelog-file.tsx]: Nitpick but the fallback value should probably be `false` to keep the logic clear.

**Comment by fregante** [comment]
Thanks @darkred! I reverted it to `false` because `undefined` clears the cache and `''` isn't a good falsey value.

**Pull request merged**